src/blockdata/transaction.rs

```diff
@@ -695,55 +695,48 @@ impl Decodable for Transaction {
-                    for txin in input.iter_mut() {
-                        txin.witness = Decodable::consensus_decode(&mut d)?;
-                    }
-                    if !input.is_empty() && input.iter().all(|input| input.witness.is_empty()) {
-                        Err(encode::Error::ParseFailed("witness flag set but no witnesses present"))
-                    } else {
-                        Ok(Transaction {
-                            version,
-                            input,
-                            output,
-                            lock_time: Decodable::consensus_decode(d)?,
-                            mw_tx: None,
-                            is_hog_ex: false
-                        })
-                    }
+
+            if segwit_flag & (1 | 8) != segwit_flag {
+                // We don't support any flag except segwit(1) and MWEB Tx(8)
+                return Err(encode::Error::UnsupportedSegwitFlag(segwit_flag))
+            }
+            
+            let mut input = Vec::<TxIn>::consensus_decode(&mut d)?;
+            let output = Vec::<TxOut>::consensus_decode(&mut d)?;
+
+            if segwit_flag & 1 != 0 { // BIP144 input witnesses
+                for txin in input.iter_mut() {
+                    txin.witness = Decodable::consensus_decode(&mut d)?;
                 }
-                8 | 9 => {
-                    let input = Vec::<TxIn>::consensus_decode(&mut d)?;
-                    let output = Vec::<TxOut>::consensus_decode(&mut d)?;
-
-                    // MimbleWimble transaction
-                    let is_mw_tx_present = u8::consensus_decode(&mut d)?;
-                    let mw_transaction: Option<mimblewimble::Transaction>  = 
-                        if is_mw_tx_present != 0 {
-                            Some(mimblewimble::Transaction::consensus_decode(&mut d)?)
-                        }
-                        else { 
-                            // HogEx transaction
-                            None
-                        };
-                    let is_hog_ex = mw_transaction.is_none();
-
-                    Ok(Transaction {
-                        version,
-                        input,
-                        output,
-                        lock_time: Decodable::consensus_decode(d)?,
-                        mw_tx: mw_transaction,
-                        is_hog_ex: is_hog_ex
-                    })
+                if !input.is_empty() && input.iter().all(|input| input.witness.is_empty()) {
+                    return Err(encode::Error::ParseFailed("witness flag set but no witnesses present"))
                 }
-                // We don't support anything else
-                x => Err(encode::Error::UnsupportedSegwitFlag(x)),
             }
+            
+            let mut mw_transaction: Option<mimblewimble::Transaction> = None;
+            let mut is_hog_ex = false;
+            if segwit_flag & 8 != 0 {
```

Comment:
@webwarrior-ws let's eliminate these magic numbers please, maybe create an int-based enum with them or just constants that explain what they mean (for example with the same rationale that David explained)